finetune/base_models/gpt/featurizer.py

```diff
@@ -242,27 +345,51 @@ def gpt_featurizer(X, encoder, config, train=False, reuse=None, explain=False, *
 
             # get the attention weights from the last layer
             if layer == config.n_layer - 1:
-                with tf.variable_scope('h%d_/h%d/attn' % (layer, layer), reuse=True):
-                    q, k, v = multihead_qkv(h, n_state=shape_list(h)[-1], n_head=config.n_heads, train=train)
+                with tf.variable_scope("h%d_/h%d/attn" % (layer, layer), reuse=True):
+                    q, k, v = multihead_qkv(
+                        h, n_state=shape_list(h)[-1], n_head=config.n_heads, train=train
+                    )
                     w = attn_weights(q, k, v, scale=True)
 
         if explain:
-            explain_out = h_out[:, initial_shape[1]:]
-            explain_out = tf.reshape(explain_out, shape=tf.concat((initial_shape[:-1], [config.n_embed]), 0))
-            h_out = h_out[:, :initial_shape[1]]
+            explain_out = h_out[:, initial_shape[1] :]
+            explain_out = tf.reshape(
+                explain_out, shape=tf.concat((initial_shape[:-1], [config.n_embed]), 0)
+            )
+            h_out = h_out[:, : initial_shape[1]]
 
         # Use hidden state at classifier token as input to final proj. + softmax
         clf_h = tf.reshape(h_out, [-1, config.n_embed])  # [batch * seq_len, embed]
-        clf_h = tf.gather(clf_h, tf.range(shape_list(X)[0], dtype=tf.int32) * config.max_length + pool_idx)
-        clf_h = tf.reshape(clf_h, shape=tf.concat((initial_shape[: -2], [config.n_embed]), 0))
-        seq_feats = tf.reshape(h, shape=tf.concat((initial_shape[:-1], [config.n_embed]), 0))
+        clf_h = tf.gather(
+            clf_h,
+            tf.range(shape_list(X)[0], dtype=tf.int32) * config.max_length + pool_idx,
+        )
+        clf_h = tf.reshape(
+            clf_h, shape=tf.concat((initial_shape[:-2], [config.n_embed]), 0)
+        )
+        seq_feats = tf.reshape(
+            h, shape=tf.concat((initial_shape[:-1], [config.n_embed]), 0)
+        )
+
+        if config.use_auxiliary_info:
+            with tf.variable_scope("context_embedding"):
+                weighted_C = tf.multiply(
+                    context, context_weighted_avg
+                )  # [batch_size, seq_length, context_dim] * [context_dim] = [batch_size, seq_length, context_dim], with weighted inputs
+                c_embed = tf.tensordot(
+                    weighted_C, context_embed_weights, axes=[[2], [0]]
+                )  # [batch_size, seq_length, context_dim] * [context_dim, n_embed] = [batch_size, seq_length, n_embed]
+                c_embed = norm(c_embed, tf.get_variable_scope())
+                seq_feats = tf.concat([seq_feats, c_embed], axis=2)
+                c_embed = tf.reduce_sum(c_embed, axis=1)
```

Comment:
Mean instead of sum.  Would also prefer if we could at least factor out this context code since it's not featurizer specific.  Don't want to have to maintain N copies of this.